lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -95,6 +95,161 @@ def self.persist_buffers!
       true #TODO: More useful return value?
     end
 
+    # Check conflicts for Configuration Type Role Updates
+    def self.conflicts_for_operables(operable_ids, crs_ids)
+      descendants_query = <<-SQL
+        WITH RECURSIVE descendants AS (
+          SELECT parent_id, child_id, parent_id as root_node FROM assignments
+          WHERE parent_id IN (#{operable_ids.join(", ")})
+          UNION ALL
+          SELECT a.parent_id, a.child_id, root_node FROM assignments a
+          INNER JOIN descendants d 
+            ON a.parent_id = d.child_id
+            WHERE a.parent_id = d.child_id
+        )
+        SELECT root_node, ra.operator_uri, crs.id FROM descendants d
+        JOIN policy_element_associations AS pea
+          ON pea.object_attribute_id = d.child_id OR pea.object_attribute_id = d.parent_id
+        JOIN policy_elements as ra
+          ON ra.id = pea.user_attribute_id
+        JOIN policy_elements as os
+          ON pea.operation_set_id = os.id
+        JOIN assignments as bb_a
+          ON os.id = bb_a.parent_id
+        JOIN assignments as crs_a
+          ON bb_a.child_id = crs_a.child_id
```

Comment:
There's a potential performance improvement by adding the `IN (#{crs_ids.join(", ")})` clause to the join here as well. 

Unfortunately, Postgres doesn't allow `IN` equality checking in joins (from what I remember, at least). You could build individual equality checks by looping over `crs_ids` and injecting each as an equality clause. Something like this may work: 
```ruby
JOIN policy_elements as crs_a 
ON os.id = bb_a.parent_id 
AND (
#{
  crs_ids.map do |crs_id| 
    "#{crs_id == crs_ids.first ? "" : "OR "}crs_a.parent_id = #{crs_id}"
  end.join(" ")
}
)
```

I'm not sure how big of a performance increase this would have, if any. If possible, we could just test it quickly on some Sandbox data or something.

Reply:
> Unfortunately, Postgres doesn't allow IN equality checking in joins (from what I remember, at least).

It does, but it's questionable whether it'll be faster or not. psql doesn't seem to like it very much if you do big `IN ()` sets.

Reply:
Yeah, I ran some quick tests and it didn't make a large performance improvement. I don't think it's necessary.